### PR TITLE
[SPARK-29349][SQL] Support FETCH_PRIOR in Thriftserver fetch request

### PR DESCRIPTION
**Pull request by juliuszsompolski**
### What changes were proposed in this pull request?

Support FETCH_PRIOR fetching in Thriftserver, and report correct fetch start offset it TFetchResultsResp.results.startRowOffset

The semantics of FETCH_PRIOR are as follow: Assuming the previous fetch returned a block of rows from offsets [10, 20)
* calling FETCH_PRIOR(maxRows=5) will scroll back and return rows [5, 10)
* calling FETCH_PRIOR(maxRows=10) again, will scroll back, but can't go earlier than 0. It will nevertheless return 10 rows, returning rows [0, 10) (overlapping with the previous fetch)
* calling FETCH_PRIOR(maxRows=4) again will again return rows starting from offset 0 - [0, 4)
* calling FETCH_NEXT(maxRows=6) after that will move the cursor forward and return rows [4, 10)

##### Client/server backwards/forwards compatibility:

Old driver with new server:
* Drivers that don't support FETCH_PRIOR will not attempt to use it
* Field TFetchResultsResp.results.startRowOffset was not set, old drivers don't depend on it.

New driver with old server
* Using an older thriftserver with FETCH_PRIOR will make the thriftserver return unsupported operation error. The driver can then recognize that it's an old server.
* Older thriftserver will return TFetchResultsResp.results.startRowOffset=0. If the client driver receives 0, it can know that it can not rely on it as correct offset. If the client driver intentionally wants to fetch from 0, it can use FETCH_FIRST.

### Why are the changes needed?

It's intended to be used to recover after connection errors. If a client lost connection during fetching (e.g. of rows [10, 20)), and wants to reconnect and continue, it could not know whether the request  got lost before reaching the server, or on the response back. When it issued another FETCH_NEXT(10) request after reconnecting, because TFetchResultsResp.results.startRowOffset was not set, it could not know if the server will return rows [10,20) (because the previous request didn't reach it) or rows [20, 30) (because it returned data from the previous request but the connection got broken on the way back). Now, with TFetchResultsResp.results.startRowOffset the client can know after reconnecting which rows it is getting, and use FETCH_PRIOR to scroll back if a fetch block was lost in transmission.

Driver should always use FETCH_PRIOR after a broken connection.
* If the Thriftserver returns unsuported operation error, the driver knows that it's an old server that doesn't support it. The driver then must error the query, as it will also not support returning the correct startRowOffset, so the driver cannot reliably guarantee if it hadn't lost any rows on the fetch cursor.
* If the driver gets a response to FETCH_PRIOR, it should also have a correctly set startRowOffset, which the driver can use to position itself back where it left off before the connection broke.
* If FETCH_NEXT was used after a broken connection on the first fetch, and returned with an startRowOffset=0, then the client driver can't know if it's 0 because it's the older server version, or if it's genuinely 0. Better to call FETCH_PRIOR, as scrolling back may anyway be possibly required after a broken connection.

This way it is implemented in a backwards/forwards compatible way, and doesn't require bumping the protocol version. FETCH_ABSOLUTE might have been better, but that would require a bigger protocol change, as there is currently no field to specify the requested absolute offset.

### Does this PR introduce any user-facing change?

ODBC/JDBC drivers connecting to Thriftserver may now implement using the FETCH_PRIOR fetch order to scroll back in query results, and check TFetchResultsResp.results.startRowOffset if their cursor position is consistent after connection errors.

### How was this patch tested?

Added tests to HiveThriftServer2Suites

### CODE REVIEW

**Comment by juliuszsompolski** [comment]
cc @wangyum @gatorsmile

**Comment by juliuszsompolski** [comment]
I did not implement it in the various metadata operations. In fact, metadata operations don't work correctly with FETCH_FIRST either right now - in e.g. https://github.com/apache/spark/blob/master/sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/operation/GetColumnsOperation.java#L231 the `extractSubset` operation deletes the rows from the start of the subset. Doing `setStartOffset(0)` on `FETCH_FIRST` after that won't work.
This bug still exists in Hive master: https://github.com/apache/hive/blob/master/service/src/java/org/apache/hive/service/cli/operation/GetColumnsOperation.java#L264
It can be fixed as a followup.

**Comment by wangyum** [comment]
@juliuszsompolski Thank you for ping me. I will review it later.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 1]
```diff
@@ -684,6 +685,94 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
       assert(e.getMessage.contains("org.apache.spark.sql.catalyst.parser.ParseException"))
     }
   }
+
+  test("ThriftCLIService FetchResults FETCH_FIRST, FETCH_NEXT, FETCH_PRIOR") {
+    def checkResult(rows: RowSet, start: Long, end: Long): Unit = {
+      assert(rows.getStartOffset() == start)
+      assert(rows.numRows() == end - start)
+      // scalastyle:off
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Please remove `// scalastyle:off`.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 2]
```diff
@@ -684,6 +685,94 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
+      rows.iterator.asScala.zip((start until end).iterator).foreach { a =>
+        val rowValue = a._1(0).asInstanceOf[Long]
+        assert(rowValue == a._2)
+      }
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Could we change this  assert to?
```scala
rows.iterator.asScala.zip((start until end).iterator).foreach { case (row, v2) =>
  assert(row(0).asInstanceOf[Long] === v2)
}
```

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 3]
```diff
@@ -684,6 +685,94 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
+    withCLIServiceClient { client =>
+      val user = System.getProperty("user.name")
+      val sessionHandle = client.openSession(user, "")
+
+      val confOverlay = new java.util.HashMap[java.lang.String, java.lang.String]
+      val operationHandle = client.executeStatement(
+        sessionHandle,
+        "SELECT * FROM range(10)",
+        confOverlay) // 10 rows result with sequence 0, 1, 2, ..., 9
+    var rows: RowSet = null
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: nit: indentation issue.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 4]
```diff
@@ -684,6 +685,92 @@ class HiveThriftBinaryServerSuite extends HiveThriftJdbcTest {
...
+      rows = client.fetchResults(
+        operationHandle, FetchOrientation.FETCH_NEXT, 5, FetchType.QUERY_OUTPUT)
+      checkResult(rows, 0, 5) // fetched [0, 5)
+
+      // Fetch another 2 rows with FETCH_NEXT
+      rows = client.fetchResults(
+        operationHandle, FetchOrientation.FETCH_NEXT, 2, FetchType.QUERY_OUTPUT)
+      checkResult(rows, 5, 7) // fetched [5, 7)
+
+      // FETCH_PRIOR 3 rows
```
Comment (wangyum) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: The `startOffset` is `5` at this time. If we do FETCH_PRIOR one row, it is 4. But PostgreSQL returns `5`. Is this what we expect?

```sql
postgres=# begin;
BEGIN
postgres=# declare rf_cur scroll cursor for select * from generate_series(0, 9);
DECLARE CURSOR
postgres=# FETCH 5 from rf_cur;
 generate_series
-----------------
               0
               1
               2
               3
               4
(5 rows)

postgres=# FETCH 2 from rf_cur;
 generate_series
-----------------
               5
               6
(2 rows)

postgres=# FETCH PRIOR from rf_cur;
 generate_series
-----------------
               5
(1 row)

postgres=# commit;
COMMIT
```

https://www.postgresql.org/docs/11/sql-fetch.html

Comment (juliuszsompolski) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: @wangyum this is expected.
`FETCH_PRIOR` of the Thriftserver is not the same as FETCH PRIOR in the cursor of the client.
Fetch in Thriftserver operates in batches of rows, and the cursor in the client caches these batches and returns results row by row. Let's say it's batching by maxRows=100, and we returned row 99. The client has rows [0, 100) from the first batch and is at row 99. The next FETCH NEXT on the cursor will have to call FETCH_NEXT to the Thriftserver to get a batch of rows [100, 200) and return row 100 to the client. Another FETCH NEXT will return row 101 from the batch without having to call FETCH_NEXT on the Thriftserver. Another FETCH NEXT on the cursor will return row 102. Then FETCH PRIOR will return row 101 again. Then FETCH PRIOR will return row 100. Only then, another FETCH PRIOR should return row 99, but the cursor doesn't have its current batch. Then it has to call FETCH_PRIOR on Thriftserver to get rows [0, 99) again.

Comment (wangyum) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Thank you @juliuszsompolski?

**Review by wangyum — APPROVED** [review]
LGTM

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
Merged to master and branch-3.0-preview.

**Comment by HyukjinKwon** [comment]
cc @jiangxb1987

**Pull request closed without merging**